lightning/src/ln/channel.rs

```diff
@@ -3554,73 +3567,83 @@ impl<Signer: Sign> Channel<Signer> {
-				//a protocol oversight, but I assume I'm just missing something.
-				if need_commitment_update {
-					if self.channel_state & (ChannelState::MonitorUpdateFailed as u32) == 0 {
-						let next_per_commitment_point = self.holder_signer.get_per_commitment_point(self.cur_holder_commitment_transaction_number, &self.secp_ctx);
-						return Ok((Some(msgs::FundingLocked {
-							channel_id: self.channel_id,
-							next_per_commitment_point,
-						}), timed_out_htlcs));
-					} else {
-						self.monitor_pending_funding_locked = true;
-						return Ok((None, timed_out_htlcs));
+				for inp in tx.input.iter() {
+					if inp.previous_output == funding_txo.into_bitcoin_outpoint() {
+						log_trace!(logger, "Detected channel-closing tx {} spending {}:{}, closing channel {}", tx.txid(), inp.previous_output.txid, inp.previous_output.vout, log_bytes!(self.channel_id()));
+						return Err(msgs::ErrorMessage {
+							channel_id: self.channel_id(),
+							data: "Commitment transaction was confirmed on chain.".to_owned()
+						});
 					}
 				}
 			}
 		}
-		Ok((None, timed_out_htlcs))
+		Ok(None)
 	}
 
-	/// Called by channelmanager based on chain blocks being disconnected.
-	/// Returns true if we need to close the channel now due to funding transaction
-	/// unconfirmation/reorg.
-	pub fn block_disconnected(&mut self, header: &BlockHeader) -> bool {
-		if self.funding_tx_confirmations > 0 {
-			self.funding_tx_confirmations -= 1;
-			if self.funding_tx_confirmations == UNCONF_THRESHOLD as u64 {
-				return true;
+	/// When a new block is connected, we check the height of the block against outbound holding
+	/// cell HTLCs in case we need to give up on them prematurely and time them out. Everything
+	/// else (e.g. commitment transaction broadcasts, channel closure detection, HTLC transaction
+	/// broadcasting, etc) is handled by the ChannelMonitor.
+	///
+	/// If we return Err, the channel may have been closed, at which point the standard
+	/// requirements apply - no calls may be made except those explicitly stated to be allowed
+	/// post-shutdown.
+	///
+	/// May return some HTLCs (and their payment_hash) which have timed out and should be failed
+	/// back.
+	pub fn update_best_block(&mut self, height: u32, highest_header_time: u32) -> Result<(Option<msgs::FundingLocked>, Vec<(HTLCSource, PaymentHash)>), msgs::ErrorMessage> {
+		let mut timed_out_htlcs = Vec::new();
+		self.holding_cell_htlc_updates.retain(|htlc_update| {
+			match htlc_update {
+				&HTLCUpdateAwaitingACK::AddHTLC { ref payment_hash, ref source, ref cltv_expiry, .. } => {
+					if *cltv_expiry <= height + HTLC_FAIL_BACK_BUFFER {
+						timed_out_htlcs.push((source.clone(), payment_hash.clone()));
+						false
+					} else { true }
+				},
+				_ => true
+			}
+		});
+
+		self.update_time_counter = cmp::max(self.update_time_counter, highest_header_time);
+
+		if self.funding_tx_confirmation_height > 0 {
```

Comment:
perhaps too short a repeated code fragment to extract, but maybe worth considering? Something like `update_confirmation_height` or similar?

Reply:
We can move it around a bit to clean it up, actually, relying on the fact that `check_get_funding_locked` will never return a funding_locked if we also want to fail the channel due to the `< minimum_depth / 2` check.